lib/push.py

```diff
@@ -83,25 +106,113 @@ def manifest_new(class_):
+      if (upload_cache is None):
+         ch.INFO("cleaning up")
+         # Delete the tarballs since we can’t yet cache them.
+         for (_, tar_c) in self.layers:
+            ch.VERBOSE("deleting tarball: %s" % tar_c)
+            tar_c.unlink_()
+
+   def json_from_path(self, path, msg):
+      "Attempt to return json from file; otherwise return None"
+      if (not path.exists()):
+         return None
+      try:
+         data = path.json_from_file(msg, raise_ok=True)
+      except:
+         return None
+      return data
+
+   def json_write(self, path, data):
+      with open(path, "w") as fp:
+         json.dump(data, fp)
+
+   def layers_from_json(self, manifest):
+      """Attempt to return a list of gzipped layer tarballs read from image
+         manifest; otherwise, return None."""
+      if (manifest is None):
+         ch.VERBOSE('manifest: None')
+         return None
+      try:
+         version = manifest['schemaVersion']
+         # load layer hashes
+         if (version == 1):
+            key1 = "fsLayers"
+            key2 = "blobSum"
+         else:  # version == 2
+            key1 = "layers"
+            key2 = "digest"
+         if (key1 not in manifest):
+            ch.VERBOSE('%s: not in manifest' % key1)
+            raise KeyError
+         layers = list()
+         for i in manifest[key1]:
+            if (key2 not in i):
+               ch.VERBOSE('%s: not in manifest' % key2)
+               raise KeyError
+            tar_p = ch.storage.upload_cache // (str(ch.digest_trim(i[key2])) + '.tar.gz')
+            if (tar_p.exists()):
+               layers.append((tar_p.file_hash(), tar_p))
+            else:
+               ch.VERBOSE('%s: does not exist' % tar_p)
+               raise KeyError
+      except KeyError:
+         ch.DEBUG("can't use previously prepared files")
+         return None
+      return layers
 
    def prepare(self):
       """Prepare self.image for pushing to self.dst_ref. Return tuple: (list
          of gzipped layer tarball paths, config as a sequence of bytes,
-         manifest as a sequence of bytes).
+         manifest as a sequence of bytes)."""
 
-         There is not currently any support for re-using any previously
-         prepared files already in the upload cache, because we don’t yet have
-         a way to know if these have changed until they are already build."""
       # Initializing an HTTP instance for the registry and doing a 'GET'
       # request right out the gate ensures the user needs to authenticate
       # before we prepare the image for upload (#1426).
       self.registry = rg.HTTP(self.dst_ref)
       self.registry.request("GET", self.registry._url_base)
+
+      ch.VERBOSE('preparing push upload files')
+      config = None
+      manifest = None
+      layers = None
+
+      if (upload_cache is not None):
+         # Check for previously prepared; if they exist, use them.
+         config = self.json_from_path(self.path_config, 'config')
```

Comment:
What's the reasoning to re-parse the stored JSON rather than just sending the existing files?

Reply:
I suppose there is no need to re-parse the stored config, however, the manifest needs to be read to get the sha256 digest for the tarball layers.